## Mobike数据集的简单评估和清理

In [1]:
#载入所需数据包
import pandas as pd
import numpy as np

In [2]:
#读入数据集
df = pd.read_csv('mobike_shanghai_sample_updated.csv')

In [3]:
#查看数据集信息
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102361 entries, 0 to 102360
Data columns (total 10 columns):
orderid             102361 non-null int64
bikeid              102361 non-null int64
userid              102361 non-null int64
start_time          102361 non-null object
start_location_x    102361 non-null float64
start_location_y    102361 non-null float64
end_time            102361 non-null object
end_location_x      102361 non-null float64
end_location_y      102361 non-null float64
track               102361 non-null object
dtypes: float64(4), int64(3), object(3)
memory usage: 6.6+ MB


In [4]:
#设置显示每列里的所有内容，而不会被默认砍掉
pd.set_option('display.max_colwidth', -1)

#查看前10行数据
df.head(10)

orderid  bikeid  userid        start_time  start_location_x  \
0  78387    158357  10080   2016-08-20 06:57  121.348            
1  891333   92776   6605    2016-08-29 19:09  121.508            
2  1106623  152045  8876    2016-08-13 16:17  121.383            
3  1389484  196259  10648   2016-08-23 21:34  121.484            
4  188537   78208   11735   2016-08-16 07:32  121.407            
5  537030   66346   10335   2016-08-07 21:00  121.488            
6  517706   99631   11258   2016-08-29 13:39  121.445            
7  270836   63136   11361   2016-08-29 09:21  121.369            
8  441362   331921  11705   2016-08-31 08:16  121.529            
9  76435    347335  8135    2016-08-30 12:49  121.467            

   start_location_y          end_time  end_location_x  end_location_y  \
0  31.389            2016-08-20 07:04  121.357         31.388           
1  31.279            2016-08-29 19:31  121.489         31.271           
2  31.254            2016-08-13 16:36  121.405         31.248           
3  31.320            2016-08-23 21:43  121.471         31.325           
4  31.292            2016-08-16 07:41  121.418         31.288           
5  31.224            2016-08-07 21:20  121.468         31.233           
6  31.176            2016-08-29 13:54  121.447         31.192           
7  31.252            2016-08-29 13:43  121.376         31.233           
8  31.324            2016-08-31 08:19  121.532         31.323           
9  31.320            2016-08-30 13:03  121.447         31.318           

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    track  
0  121.347,31.392#121.348,31.389#121.349,31.390#121.350,31.390#121.351,31.390#121.351,31.391#121.352,31.391#121.353,31.391#121.353,31.394#121.354,31.391#121.355,31.391#121.356,31.389#121.356,31.390#121.356,31.391#121.357,31.388#121.357,31.389                                                                                                                                                                                                                                                                                                                                                                                         
1  121.489,31.270#121.489,31.271#121.490,31.270#121.490,31.271#121.491,31.271#121.492,31.270#121.492,31.271#121.493,31.269#121.493,31.270#121.494,31.268#121.494,31.269#121.495,31.269#121.495,31.270#121.496,31.270#121.497,31.270#121.497,31.271#121.498,31.272#121.499,31.272#121.499,31.273#121.500,31.273#121.500,31.274#121.501,31.274#121.502,31.273#121.502,31.274#121.503,31.273#121.504,31.273#121.504,31.274#121.505,31.274#121.505,31.275#121.506,31.275#121.506,31.276#121.506,31.277#121.507,31.277#121.507,31.278#121.507,31.279#121.508,31.279                                                                             
2  121.381,31.251#121.382,31.251#121.382,31.252#121.382,31.253#121.382,31.254#121.383,31.250#121.383,31.251#121.383,31.254#121.384,31.250#121.385,31.250#121.386,31.250#121.387,31.250#121.387,31.251#121.388,31.251#121.388,31.252#121.389,31.251#121.389,31.252#121.390,31.251#121.391,31.251#121.392,31.251#121.393,31.248#121.393,31.250#121.394,31.248#121.394,31.249#121.394,31.250#121.395,31.248#121.396,31.248#121.397,31.248#121.398,31.248#121.399,31.248#121.400,31.248#121.401,31.248#121.402,31.248#121.403,31.248#121.404,31.248#121.404,31.249#121.405,31.249#121.406,31.248#121.406,31.249#121.407,31.248#121.407,3

In [5]:
#检查数据集是否有重复订单
df.orderid.duplicated().sum()

0

In [6]:
#查看总用户数
df.userid.nunique()

16887

In [7]:
#查看总单车数
df.bikeid.nunique()

79062

以上可见，数据集涉及到102361条订单数据，16887个用户和79062辆单车。

由于骑行轨迹列，对我提出的5个问题的探究没有太多帮助，以下删除该列。

In [8]:
#删除骑行轨迹列
df = df.drop(['track'],axis=1)

In [9]:
#将start_time和end_time列转成正确的时间格式，以便后续进行计算
df.start_time = pd.to_datetime(df.start_time)
df.end_time = pd.to_datetime(df.end_time)

In [10]:
#计算骑行时长，存入新列duration中
df['duration'] = df.end_time - df.start_time
df['duration']

0        00:07:00
1        00:22:00
2        00:19:00
3        00:09:00
4        00:09:00
5        00:20:00
6        00:15:00
7        04:22:00
8        00:03:00
9        00:14:00
10       00:08:00
11       00:46:00
12       00:11:00
13       00:07:00
14       00:13:00
15       00:17:00
16       00:11:00
17       00:10:00
18       00:05:00
19       00:13:00
20       00:41:00
21       00:03:00
22       00:11:00
23       00:06:00
24       00:22:00
25       00:07:00
26       00:09:00
27       00:06:00
28       00:08:00
29       00:29:00
           ...   
102331   00:06:00
102332   00:04:00
102333   00:08:00
102334   00:51:00
102335   00:06:00
102336   00:12:00
102337   00:35:00
102338   00:05:00
102339   00:08:00
102340   00:07:00
102341   00:08:00
102342   00:14:00
102343   00:08:00
102344   00:07:00
102345   00:07:00
102346   00:02:00
102347   00:21:00
102348   00:10:00
102349   00:38:00
102350   00:06:00
102351   00:14:00
102352   00:04:00
102353   00:07:00
102354   00:08:00
102355   0

In [11]:
#将duration列的时间格式转化成分钟数
df['duration'] = df['duration'].apply(lambda x: x/np.timedelta64(1,'m'))

In [12]:
#根据start_time排序
df.sort_values('start_time')

orderid  bikeid  userid          start_time  start_location_x  \
91307   31355    79699   759    2016-08-01 00:23:00  121.520            
15161   31388    33279   1440   2016-08-01 00:25:00  121.505            
67939   31401    3060    5463   2016-08-01 00:26:00  121.375            
94977   31423    43119   4666   2016-08-01 00:32:00  121.528            
47303   31438    116238  2193   2016-08-01 00:33:00  121.434            
18021   31455    115749  5114   2016-08-01 00:35:00  121.436            
59375   31480    61084   1183   2016-08-01 00:43:00  121.387            
67326   31485    123910  2154   2016-08-01 00:44:00  121.495            
1146    31502    36158   6667   2016-08-01 00:45:00  121.528            
31506   31508    26060   4756   2016-08-01 00:46:00  121.424            
47051   31540    112334  7066   2016-08-01 00:53:00  121.427            
57541   31549    101533  3364   2016-08-01 00:54:00  121.379            
3001    31564    101807  1383   2016-08-01 00:58:00  121.546            
64005   31572    104356  3196   2016-08-01 01:00:00  121.386            
53306   31581    64570   4622   2016-08-01 01:02:00  121.534            
21805   31590    113094  3250   2016-08-01 01:06:00  121.538            
90613   31613    26116   5069   2016-08-01 01:11:00  121.475            
24027   31606    121853  5432   2016-08-01 01:11:00  121.453            
50690   31619    823     6615   2016-08-01 01:13:00  121.473            
68825   31633    41445   6172   2016-08-01 01:18:00  121.398            
66030   31635    50954   2077   2016-08-01 01:19:00  121.436            
14593   31645    78767   6378   2016-08-01 01:21:00  121.518            
4902    31658    104609  6665   2016-08-01 01:25:00  121.539            
69234   31708    9330    7636   2016-08-01 01:42:00  121.429            
6917    31707    113275  3084   2016-08-01 01:43:00  121.411            
45709   31740    124887  4161   2016-08-01 01:51:00  121.459            
87228   31742    108520  1820   2016-08-01 01:51:00  121.449            
60663   31771    98978   1312   2016-08-01 02:02:00  121.353            
77013   31798    102468  77     2016-08-01 02:16:00  121.545            
44620   31807    13832   5553   2016-08-01 02:18:00  121.424            
...       ...      ...    ...                   ...      ...            
39692   1737723  136178  16452  2016-08-31 23:39:00  121.506            
71025   1737731  359079  17377  2016-08-31 23:39:00  121.507            
60311   1737737  193007  710    2016-08-31 23:39:00  121.460            
93284   1737747  340068  13470  2016-08-31 23:40:00  121.387            
29299   1737753  156394  1008   2016-08-31 23:40:00  121.498            
68334   1737764  234952  4562   2016-08-31 23:41:00  121.467            
97664   1737783  251330  2017   2016-08-31 23:42:00  121.527            
72056   1737809  245129  698    2016-08-31 23:43:00  121.561            
67612   1737859  167246  12929  2016-08-31 23:45:00  121.386            
44012   1737891  152062  9808   2016-08-31 23:46:00  121.407            
67087   1737893  167397  11526  2016-08-31 23:46:00  121.375            
77489   1737884  387349  10135  2016-08-31 23:46:00  121.484            
1908    1737902  192206  8529   2016-08-31 23:47:00  121.471            
49193   1737942  387351  8851   2016-08-31 23:49:00  121.428            
99359   1737960  358810  14051  2016-08-31 23:50:00  121.589            
3278    1737988  224756  11446  2016-08-31 23:51:00  121.412            
18828   1737981  267558  7303   2016-08-31 23:51:00  121.507            
13624   1737991  387162  14714  2016-08-31 23:51:00  121.443            
84373   1737998  300869  12015  2016-08-31 23:52:00  121.390            
30965   1738005  263389  12131  2016-08-31 23:52:00  121.442            
79018   1738032  379992  3631   2016-08-31 23:53:00  121.467            
2562    1738038  215704  9716   2016-08-31 23:53:00  121.367            
45090   1738023  243092  10839  2016-08-31 23:53:00  121.540      

以上可以看出数据集记录了2016年8月1日到8月31日的订单，即8月整月的数据。

In [13]:
#查看用户骑行次数
df.userid.value_counts()

6199     25
3369     23
7475     21
6138     21
5150     21
5949     21
8633     21
6526     21
4123     21
7883     20
8762     20
11456    20
5215     20
10609    20
8700     20
1123     20
9708     19
7045     19
6450     19
7758     19
1866     19
6544     19
9604     19
9188     19
8188     19
7776     19
8077     19
7376     19
11258    18
3175     18
         ..
16397    1 
14217    1 
4167     1 
16461    1 
306      1 
16438    1 
16792    1 
34       1 
9856     1 
10154    1 
17073    1 
8886     1 
13206    1 
15191    1 
16490    1 
4015     1 
17137    1 
16554    1 
16003    1 
418      1 
1668     1 
16984    1 
14312    1 
16954    1 
14018    1 
16682    1 
16856    1 
17201    1 
16131    1 
5295     1 
Name: userid, Length: 16887, dtype: int64

In [14]:
#对骑行时长进行排序
df.duration.sort_values(ascending=False)

61444     4725.0
12744     3662.0
85894     2947.0
50219     2848.0
36431     2479.0
48841     1961.0
67143     1538.0
53291     1511.0
87553     1393.0
68797     1272.0
36612     1227.0
79854     1115.0
51671     1072.0
57505     1053.0
57343     955.0 
74862     917.0 
96400     905.0 
32219     892.0 
37656     889.0 
14287     840.0 
100088    666.0 
38982     642.0 
27415     629.0 
96043     627.0 
101327    604.0 
35991     604.0 
42508     602.0 
14580     573.0 
71945     564.0 
44430     530.0 
          ...   
49950     1.0   
6589      1.0   
83471     1.0   
63854     1.0   
89778     1.0   
29935     1.0   
25869     1.0   
39787     1.0   
82616     1.0   
14890     1.0   
29321     1.0   
98446     1.0   
49560     1.0   
11605     1.0   
81845     1.0   
3719      1.0   
52106     1.0   
30241     1.0   
26732     1.0   
78308     1.0   
86453     1.0   
89182     1.0   
44415     1.0   
5538      1.0   
100957    1.0   
75065     1.0   
54773     1.0   
91480     1.0 

In [15]:
#查看骑行时长最长的数据
df.iloc[61444]

orderid             119909             
bikeid              86648              
userid              7426               
start_time          2016-08-05 08:43:00
start_location_x    121.507            
start_location_y    31.258             
end_time            2016-08-08 15:28:00
end_location_x      121.503            
end_location_y      31.265             
duration            4725               
Name: 61444, dtype: object

以上数据可见，起始坐标基本没变，骑行时间确持续了3天多。明显为没有锁车导致的。

In [16]:
#查看另外一个骑行时长较长的行
df.iloc[36431]

orderid             1796813            
bikeid              109837             
userid              13933              
start_time          2016-08-26 21:10:00
start_location_x    121.383            
start_location_y    31.247             
end_time            2016-08-28 14:29:00
end_location_x      121.39             
end_location_y      31.258             
duration            2479               
Name: 36431, dtype: object

以上行也是同样的情况，应该是没有锁车的异常情况。

此数据集中有不少类似的异常情况，因此在可视化的过程中，需要考虑到针对此类数据的统计。

In [17]:
#再次查看整理后的数据
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102361 entries, 0 to 102360
Data columns (total 10 columns):
orderid             102361 non-null int64
bikeid              102361 non-null int64
userid              102361 non-null int64
start_time          102361 non-null datetime64[ns]
start_location_x    102361 non-null float64
start_location_y    102361 non-null float64
end_time            102361 non-null datetime64[ns]
end_location_x      102361 non-null float64
end_location_y      102361 non-null float64
duration            102361 non-null float64
dtypes: datetime64[ns](2), float64(5), int64(3)
memory usage: 7.8 MB


In [18]:
#将整理过的数据集存为“mobike_shanghai_clean.csv”
df.to_csv('mobike_shanghai_clean.csv')

整理过的数据集就可以导入Tableau，用于可视化探究了。